tests/Rules.Framework.BenchmarkTests/Tests/Benchmark3/Benchmark3.cs

```diff
@@ -8,16 +8,16 @@ namespace Rules.Framework.BenchmarkTests.Tests.Benchmark3
     [SkewnessColumn, KurtosisColumn]
     public class Benchmark3 : IBenchmark
     {
-        private readonly Benchmark3Data benchmarkData = new Benchmark3Data();
-        private RulesEngine<ContentTypes, ConditionTypes> rulesEngine;
+        private readonly Scenario8Data benchmarkData = new Scenario8Data();
+        private RulesEngine<ContentTypes, ConditionTypes>? rulesEngine;
 
         [ParamsAllValues]
         public bool EnableCompilation { get; set; }
 
         [Benchmark]
         public async Task RunAsync()
         {
-            await this.rulesEngine.MatchOneAsync(ContentTypes.TexasHoldemPokerSingleCombinations, this.benchmarkData.MatchDate, this.benchmarkData.Conditions).ConfigureAwait(false);
+            await this.rulesEngine!.MatchOneAsync(ContentTypes.TexasHoldemPokerSingleCombinations, this.benchmarkData.MatchDate, this.benchmarkData.Conditions).ConfigureAwait(false);
```

Comment:
I didn't know that operator `!.` 🙂

Reply:
It is the new "null forgiving operator". As you may already know, C# 9.0 onwards introduced the "nullable operator" for all types. Up until then, you'd only apply the nullable operator to value types (`int`, `DateTime`, an `enum` defined by you, etc), and since then also reference types have supported this operator.

When you create a new `class` definition, with a property of a type whose default value is _potentially_ `null`, the compiler reports to you a warning to either define the initial value at the constructor of that `class` or mark the property as nullable. If you choose to mark it as nullable, all references to `get` the value of that property will report a warning stating "value might be null here". At those references, you analyze the code, and come to the conclusion that the property at that specific spot can never be `null`, so you add the null forgiving operator (`!`) so that the compiler knows that you have analyzed that code and the value of the property can never be `null` at that point.

So, bottom line, the null forgiving operator is a kind of "sugar coating" on top of your code, it has no side effect at all on your implementation (code works all the same if you resolve the warnings or not) but it lets the compiler know that you have revised that code and everything is OK. It keeps the developers with a watchful eye for `NullReferenceException` 👀